# scRNASeq
Repository for CCBR single cell RNA-Seq pipeline

This is a mirrored repository of the live CCBR Pipeliner implementation of the scRNASeq pipeline. The CCBR Pipeliner can be cloned from here:
github.com/CCBR/Pipeliner

This repository currently contains R scripts that generate the initial and QC reports, principal components visualization, clustering, genelists, and tSNE plots. The workflow for these graphics and files are split up into three seperate Rmarkdown scripts that each render an html report.

Currently, the initial qc report details info about outlier cells and lowly expressed genes that are filtered out. It also provides visualizations of principal components that can then be used to determine how many PCs to use to cluster.

Data processing and filtering is done using the scater package. Cell cycle stage analysis is done in the initial report using SCRAN cyclone. The data is then imported into Seurat for principal components analysis and PC visualization.

The PC visualization in the initial report takes the form of a top gene scores for each compoenet generated by VizPCA, heatmaps of the extreme cells for the top genes of each component generated by PCHeatmap, and an elbow plot showing the standard deviation of each component generated by PCElbowPlot.

The initial report also contains a heatmap of cells by GSVA enrichment score for KEGG pathways.

The Jackstraw report generates Jackstraw plots for each component which is another helpful determination of significant PCs.

The clustering report takes two important parameters, the resolution (which is used by Seurat to determine the number of clusters) and the max number of PCs to use (i.e. 1 through the specified number of PCs will be used for clustering of cells). Seurat's FindCluster algorithm is then run to generate cluster identities and then a tSNE plot is generated. Genelists are provided for differentially expressed genes of each cluster, in the form of a report based list of the top 100 genes, and text files that contain all genes that meet the significance threshold for each cluster. The clustering report and genelist text files are prefixed by the number of components and resolution that is selected.

All Rmarkdown files are called using the call Rscript with specified parameters.

Future directions include:
- cluster by cluster heatmaps using GSVA pathway enrichment after the clustering step
- displaying more PCs in the visualization steps
- more filtering options


Current workflow flowchart:

![alt text](scrnaseq_flowchart.png?raw=true)
